File: algorithms/2_medium/M/02087_minimum_cost_homecoming_of_a_robot_in_a_grid/solution.cpp
class Solution {
public:
    static int minCost(
        const vector<int>& start_pos,
        const vector<int>& home_pos,
        const vector<int>& row_costs,
        const vector<int>& col_costs
    ) {
        int cost = 0;
        int x = start_pos[0];
        int y = start_pos[1];
        int hx = home_pos[0];
        int hy = home_pos[1];

        while (x != hx) {
            x += (x < hx ? 1 : -1);
            cost += row_costs[x];
        }

        while (y != hy) {
            y += (y < hy ? 1 : -1);
            cost += col_costs[y];
        }

        return cost;
    }
};

File: algorithms/2_medium/M/02087_minimum_cost_homecoming_of_a_robot_in_a_grid/solution.rs
impl Solution {
    pub fn min_cost(start_pos: Vec<i32>, home_pos: Vec<i32>, row_costs: Vec<i32>, col_costs: Vec<i32>) -> i32 {
        let mut cost = 0_i32;
        let (mut x, mut y) = (start_pos[0], start_pos[1]);
        let (hx, hy) = (home_pos[0], home_pos[1]);

        while x != hx {
            x += if x < hx { 1 } else { -1 };
            cost += row_costs[x as usize];
        }

        while y != hy {
            y += if y < hy { 1 } else { -1 };
            cost += col_costs[y as usize];
        }

        cost
    }
}
